src/libraries/System.Console/tests/KeyParserTests.cs

```diff
@@ -0,0 +1,1715 @@
+    [MemberData(nameof(AsciiCharactersArguments))]
+    public void VeraseIsRespected(TerminalData terminalData, char input, ConsoleKey ifNotVerase)
+    {
+        ConsoleKeyInfo consoleKeyInfo = Map(new[] { input }, terminalData.TerminalDb, (byte)input, 1);
+
+        Assert.Equal(input, consoleKeyInfo.KeyChar);
+        Assert.Equal(ConsoleKey.Backspace, consoleKeyInfo.Key);
+        Assert.NotEqual(ifNotVerase, consoleKeyInfo.Key);
+        Assert.Equal((ConsoleModifiers)0, consoleKeyInfo.Modifiers);
+    }
+
+    [Fact]
+    public void NewLineEscapeSequenceProducesCharacter()
+    {
+        XTermData xTerm = new();
+
+        ConsoleKeyInfo consoleKeyInfo = Map("\u001BOM".ToCharArray(), xTerm.TerminalDb, xTerm.Verase, 3);
+
+        Assert.Equal(ConsoleKey.Enter, consoleKeyInfo.Key);
+        Assert.Equal('\r', consoleKeyInfo.KeyChar);
+        Assert.Equal((ConsoleModifiers)0, consoleKeyInfo.Modifiers);
+    }
+
+    [Fact]
+    public void BackTabEscapeSequence()
+    {
+        XTermData xTerm = new();
+
+        ConsoleKeyInfo consoleKeyInfo = Map("\u001B[Z".ToCharArray(), xTerm.TerminalDb, xTerm.Verase, 3);
+
+        Assert.Equal(ConsoleKey.Tab, consoleKeyInfo.Key);
+        Assert.Equal(default, consoleKeyInfo.KeyChar);
+        Assert.Equal(ConsoleModifiers.Shift, consoleKeyInfo.Modifiers);
+    }
+
+    private static ConsoleKeyInfo Map(char[] chars, ConsolePal.TerminalFormatStrings terminalFormatStrings, byte verase, int expectedStartIndex)
+    {
+        int startIndex = 0;
+
+        KeyParser.Parse(chars, terminalFormatStrings, 0, verase,
+            out ConsoleKey consoleKey, out char ch, out bool isShift, out bool isAlt, out bool isCtrl, ref startIndex, chars.Length);
+        //Assert.True(Net6KeyParser.Parse(chars, terminalFormatStrings, 0, verase,
+        //    out ConsoleKey consoleKey, out char ch, out bool isShift, out bool isAlt, out bool isCtrl, ref startIndex, chars.Length));
+
+        Assert.Equal(expectedStartIndex, startIndex);
+
+        return new ConsoleKeyInfo(ch, consoleKey, isShift, isAlt, isCtrl);
+    }
+}
+
+public abstract class TerminalData
+{
+    private ConsolePal.TerminalFormatStrings? _terminalDb;
+    private Encoding? _consoleEncoding;
+
+    protected abstract string EncodingCharset { get; }
+    protected abstract string Term { get; }
+    protected abstract string EncodedTerminalDb { get; }
+    internal abstract byte Verase { get; }
+    internal abstract IEnumerable<(byte[], ConsoleKeyInfo)> RecordedScenarios { get; }
+
+    internal ConsolePal.TerminalFormatStrings TerminalDb => _terminalDb ??=
+        new ConsolePal.TerminalFormatStrings(new TermInfo.Database(Term, Convert.FromBase64String(EncodedTerminalDb)));
+
+    internal Encoding ConsoleEncoding => _consoleEncoding ??= (string.IsNullOrEmpty(EncodingCharset) ? Encoding.Default : Encoding.GetEncoding(EncodingCharset)).RemovePreamble();
+}
+
+// Below you can find test data recorded with https://github.com/adamsitnik/ReadKey
```

Comment:
Not a problem that needs solving to merge this, but I wonder whether there is a place we can put this tool that's in dotnet/*. If somehow something happened to your repo, these tests would be unmaintainable.

I wonder whether dotnet/runtime-assets might be convenient. It's already set up with arcade etc to build, this would just be something that didn't ever get packaged. The build should be set up to only kick off when the repo changes, so it's low overhead but verifies it doesnt rot.